doc/reference.md

```diff
@@ -0,0 +1,509 @@
+
+```
+[ 
+  [x * 2, y]
+  for x in [1, 2, 3, 4, 5]
+  for y in [1, 2, 3]
+  if x % 2 == 0
+]
+```
+
+```
+std.flatMap(
+  function(x) std.flatMap(
+    function(y) if x % 2 == 0 then [[x * 2, y]] else [],
+    [1, 2, 3]
+  ),
+  [1, 2, 3, 4, 5]
+)
+```
+
+Arrays can be concatenated using the operator `+`. Arrays `a` and `b` are equal if they have equal length and for all indexes `i`, `a[i] == b[i]`. The comparison of arrays is "lexicographic", so array `a` is smaller than `b` if `a[i] < b[i]` for some `i` and for all `j < i`, `a[i] == b[i]`. 
+
+### Object
+
+In the simplest case Jsonnet object is a mapping from string keys to arbitrary values.
+
+```
+{
+  "foo": 1,
+  "bar": {
+    "arr": [1, 2, 3],
+    "number": 10 + 7,
+  }
+}
+```
+
+Jsonnet objects can be indexed either using `.` and an identifier or `[]` with an arbitrary expression.
+
+```
+local obj = {
+  "foo": 1,
+  "bar": {
+    "arr": [1, 2, 3],
+    "number": 10 + 7,
+  }
+};
+[
+  obj.foo,
+  obj["foo"],
+  obj["f" + "oo"]
+]
+```
+
+### Inheritance
+
+Jsonnet objects allow inheritance in the OOP sense, even though there are no classes or declarations. The inheritance is realized as an operation `+` which can be applied to any two objects. This might be surprising, because in mainstream languages the inheritance hierarchy is static.
+
+For objects which are simple key-value mappings, inheritance is the same as replacing the respective fields from the first object with fields from the second object.
+
+```
+{
+  a: 1
+  b: 2,
+}
++
+{
+  a: 3
+}
+```
+
+It is possible for one field to refer to other ones using `self` and it is possible to refer to the fields of the base object using `super`.
+
+```
+local obj = {
+  name: "Alice",
+  greeting: "Hello, " + self.name,
+};
+[
+  obj,
+  obj + { name: "Bob" },
+  obj + { greeting: super.greeting + "!"}
+]
+```
+
+In general, it is useful to think about Jsonnet object as a stack of *layers*. A layer consists of fields. An object literal or an object comprehension is a single-layer object. Object inheritance `A` + `B` results in an object with `B` layers put on top of `A` layers. 
+
+Reference to a field through `self` corresponds to looking for the field starting from the top of the stack and going towards the bottom until the field is found. Reference through `super` searches for the field starting from the layer below the current one.
+
+It is easy to emulate functions using objects. It goes without saying that it is not a good style to use in the real code. 
+
+```
+local add = {
+  params: {
+    a: error "please provide argument a",
+    b: error "please provide argument b",
+  },
+  result: self.params.a + self.params.b
+};
+(add + { params: { a: 1, b: 2} }).result
+```
+
+#### Properties
+
+Let `D`, `E`, `F` range over arbitrary expressions evaluating to objects. Let ≡ mean equivalence.
+
+* **Associativity** always holds:
```

Comment:
This is from the spec is it not?  I'm ambivilent about moving it here, but I don't think we should say it twice.

Also, I think it's a good idea to move the lexing part of the spec into here, because it was never really formally defined in the first place.  The spec can simply say there is no formal definition of lexing and to read the reference manual to understand it.

Reply:
This used to be a part of the spec, but you moved it out during the visual revamp of the website. This is not really a part of the spec, but a consequence of the spec. I mean these are properties that you can prove based on the spec. I think the reference is the right place for properties like that.

I'm not convinced about the lexing. It is not as formal as the rest of the stuff in the spec, but it is necessary to write a compliant implementation. A user doesn't really care about the lexing as a separate stage. Also, I think it is pretty obvious and conventional. There is some stuff from there that I would like to have here is the description of different kinds of strings and different kinds of comments.